docs/source/designs/backup-recovery.rst

```diff
@@ -0,0 +1,515 @@
+At any point if the diffs grow to large the process can be started over with a new, recent copy of the database.
+
+Text Deltas
+~~~~~~~~~~~
+
+The full contents of a SLQite3 database can be dumped as SQL text at any time.
+The *Binary Deltas* design could be applied to these SQL text dumps instead.
+Text diffs could be compressed to reduce the overhead compared to binary deltas.
+These diffs are likely to be slightly easier to work with in the event any problems arise.
+
+*What we've considered.*
+*What trade-offs are involved with each choice.*
+*Why we've chosen the one we did.*
+
+Detailed Implementation Design
+------------------------------
+
+"Application SQL Log" is the chosen design.
+
+*Focus on:*
+
+* external and internal interfaces
+* how externally-triggered system events (e.g. sudden reboot; network congestion) will affect the system
+* scalability and performance
+
+State
+~~~~~
+
+A replica consists of the following pieces of state:
+
+#. a snapshot
+
+   A snapshot the minimal sequence of SQL statements
+   (with arguments embedded)
+   which will re-create the database from which it was created [10]_.
+   A snapshot can be produced using the ``iterdump`` feature of the Python SQLite3 module.
+
+#. an event stream
+
+   An event stream is a sequence of SQL statements
+   (with arguments embedded)
+   which update a certain database snapshot.
+   Each statements in the event stream is paired with a sequence number.
+   Sequence numbers are assigned in a monotonically increasing sequence that corresponds to the order in which the statements were executed.
+   These can be used to address a specific statement within the event stream.
+
+#. a sequence number
+
+   A snapshot includes state which was created by statements from some prefix of the event stream.
+   The snapshot is paired with a sequence number indicating the last statement in this prefix.
+   This allows recovery to find the correct position in the event stream to begin replaying statements to reconstruct the newest version of the database.
+
+The event stream is represented in the local database in a new table::
+
+  CREATE TABLE [event-stream] (
+      -- A sequence number which allows us to identify specific positions in
+      -- the sequence of modifications which were made to the database.
+      [sequence-number] INTEGER PRIMARY KEY,
+
+      -- A SQL statement which likely made a change to the database state.
+      [statement] TEXT,
+  );
+
+Arguments are substituted into the statement so that they match the form of statements generated during the *snapshot* phase.
+
+Replication
+~~~~~~~~~~~
+
+The replication process is as follows:
+
+#. Replication is configured using the external interface.
+
+   #. The *replica directory*,
+      a new mutable directory,
+      is created on grid.
+   #. The write capability is added to the database.
+   #. The read capability is returned to the external caller.
```

Comment:
Also, if the writecap is contained in the database and that database is later restored, does that mean that the newly-restored instance will (potentially) be writing to the same (replica?) directory as the old device/instance? Can/should additional care be taken to prevent old possessors of the writecap from messing with new/post-restore replicas/snapshots?

Reply:
> I'm not sure what is achieved by diminishing this cap. If a) the writecap is in the (remotely-stored?) database and b) the readcap points to that database, then the possessor of the readcap can trivially discover the writecap.

Hmm.  Indeed.  I thought to diminish it just because no one outside of the system ever needs the write cap but it is true that this readcap can be leveraged to obtain the writecap...

I'm not sure what to do about that.  Maybe it is still a good idea to diminish it and then try to figure out a way to remove the property that allows an attacker to leverage it into a writecap (maybe later if we can't figure out how to do it now)?  Put another way, "there is a hole over there so we might as well make a same-shaped hole over here" doesn't make sense, even though having one hole is still bad.

> Also, if the writecap is contained in the database and that database is later restored, does that mean that the newly-restored instance will (potentially) be writing to the same (replica?) directory as the old device/instance?

Hm.  I considered this a feature of the design, not a flaw.  But maybe that was a mistake.

If we want the newly-restored instance to use a different replica we can arrange that, I think.  For example, we could say that a newly-restored instance forgets its replication configuration and must be re-configured.  Generate a new replica dircap, re-upload (maybe if we're really careful the "re-upload" just gives you the same immutable caps but in a new location, or maybe we can take explicit steps to make sure that is what happens ... but that's just an optimization), proceed to replicate to the new location, the old one is forgotten.

Maybe this even suggests that the replica writecap is either not kept in the database or is not replicated along with the rest of the database state - which closes the above-mentioned hole, too.

Reply:
At least returning a read-only cap to the caller makes it more-obvious that they shouldn't be writing to it .. leaking the read-cap is still bad (and I guess leaking the write-cap is _even worse_ since you could then wipe the data).

> For example, we could say that a newly-restored instance forgets its replication configuration and must be re-configured.

I think this is the right thing, to avoid the situation where the "old device" is found/used again (and then you have multiple writers to the same writecap).